contributors/design-proposals/storage/container-storage-interface.md

```diff
@@ -0,0 +1,434 @@
+
+#### Internal Interfaces
+
+The in-tree CSI volume plugin will implement the following internal Kubernetes volume interfaces:
+
+1. `VolumePlugin`
+    * Mounting/Unmounting of a volume to a specific path.
+2. `AttachableVolumePlugin`
+    * Attach/detach of a volume to a given node.
+
+Notably, `ProvisionableVolumePlugin` and `DeletableVolumePlugin` are not implemented because provisioning and deleting for CSI volumes is handled by an external provisioner.
+
+#### Mount and Unmount
+
+The in-tree volume plugin’s SetUp and TearDown methods will trigger the `NodePublishVolume` and `NodeUnpublishVolume` CSI calls via Unix Domain Socket. Kubernetes will generate a unique `target_path` (unique per pod per volume) to pass via `NodePublishVolume` for the CSI plugin to mount the volume. Upon successful completion of the `NodeUnpublishVolume` call (once volume unmount has been verified), Kubernetes will delete the directory.
+
+The Kubernetes volume sub-system does not currently support block volumes (only file), so for alpha, the Kubernetes CSI volume plugin will only support file.
+
+As part of kubelet work, NodeAuthorizer needs to be updated to allow kubelet accessing MountSecret for internal CSI volume plugin.
+
+#### Attaching and Detaching
+
+The attach/detach controller,running as part of the kube-controller-manager binary on the master, decides when a CSI volume must be attached or detached from a particular node.
+
+When the controller decides to attach a CSI volume, it will call the in-tree CSI volume plugin’s attach method. The in-tree CSI volume plugin’s attach method will do the following:
+
+1. Create a new `VolumeAttachment` object (defined in the “Communication Channels” section) to attach the volume.
+    * The name of the of the `VolumeAttachment` object will be the `<UniqueVolumeName>-<NodeId>`.
+      * `UniqueVolumeName` is `<PluginName>/<VolumeName>`
+        * `VolumeName` is the name/ID uniquely identifying the actual backing device, directory, path for a particular volume.
+        * `PluginName` is `CSI/<DriverName>`
+      * `NodeId` is the ID of the node the volume should be attached to.
+    * If a `VolumeAttachment` object with the corresponding name already exists, the in-tree volume plugin will simply begin to poll it as defined below. The object is not modified; only the external-attacher should change the status fields; and the external-attacher is responsible for it’s own retry and error handling logic.
+2. Poll the `VolumeAttachment` object waiting for one of the following conditions:
+    * The `VolumeAttachment.Status.IsAttached` field to become `true`.
+      * The operation completes successfully.
+    * An error to be set in the `VolumeAttachment.Status.AttachError` field.
+      * The operation terminates with the specified error.
+    * The operation to timeout.
+      * The operation terminates with timeout error.
```

Comment:
Should clarify how to recover the volume if the external attacher did actually attach the volume but perhaps was delayed. Will it be like a subsequent attach to the same node succeeds because the external attacher will respond immediately and set _IsAttached_ to _true_?

What happens if on timeout Kubernetes decides to attach the same volume on another node? While the volume is attached to the other node but Kubernetes timedout?

Reply:
Kubernetes A/D controller will try to attach the volume periodically until corresponding pod exists. I.e. it won't delete VolumeAttachment instance and external attacher will keep trying.

When the pod is deleted, A/D controller detaches the volume, i.e. deletes VolumeAttachment. That's signal for the external attacher to detach the volume.